examples/ensemble/plot_gradient_boosting_quantile.py

```diff
@@ -3,77 +3,268 @@
+#
+# Tuning the hyper-parameters of the quantile regressors
+# ------------------------------------------------------
+#
+# In the plot above, we observed that the 5th percentile regressor seems to
+# underfit and could not adapt to sinusoidal shape of the signal.
+#
+# The hyper-parameters of the model were approximately hand-tuned for the
+# median regressor and there is no reason than the same hyper-parameters are
+# suitable for the 5th percentile regressor.
+#
+# To confirm this hypothesis, we tuned the hyper-parameters of a new regressor
+# of the 5th percentile by selecting the best model parameters by
+# cross-validation on the pinball loss with alpha=0.05:
+
+# %%
+from sklearn.model_selection import RandomizedSearchCV
+from sklearn.metrics import make_scorer
+from pprint import pprint
+
+
+param_grid = dict(
+    learning_rate=[0.01, 0.05, 0.1],
+    n_estimators=[100, 150, 200, 250, 300],
+    max_depth=[2, 5, 10, 15, 20],
+    min_samples_leaf=[1, 5, 10, 20, 30, 50],
+    min_samples_split=[2, 5, 10, 20, 30, 50],
+)
+alpha = 0.05
+neg_pinball_loss_05p_scorer = make_scorer(
+    pinball_loss,
+    alpha=alpha,
+    greater_is_better=False,  # maximize the negative loss
+)
+gbr = GradientBoostingRegressor(loss="quantile", alpha=alpha, random_state=0)
+search_05p = RandomizedSearchCV(
+    gbr,
+    param_grid,
+    n_iter=10,  # increase this if computational budget allows
+    scoring=neg_pinball_loss_05p_scorer,
+    n_jobs=2,
+    random_state=0,
+).fit(X_train, y_train)
+pprint(search_05p.best_params_)
+
+# %%
+# We observe that the search procedure identifies that deeper trees are needed
+# to get a good fit for the 5th percentile regressor. Deeper trees are more
+# expressive and less likely to underfit.
+#
+# Let's now tune the hyper-parameters for the 95th percentile regressor. We
+# need to redefine the `scoring` metric used to select the best model, along
+# with adjusting the alpha parameter of the inner gradient boosting estimator
+# itself:
+from sklearn.base import clone
+
+alpha = 0.95
+neg_pinball_loss_95p_scorer = make_scorer(
+    pinball_loss,
+    alpha=alpha,
+    greater_is_better=False,  # maximize the negative loss
+)
+search_95p = clone(search_05p).set_params(
+    estimator__alpha=alpha,
+    scoring=neg_pinball_loss_95p_scorer,
+)
+search_95p.fit(X_train, y_train)
+pprint(search_95p.best_params_)
 
-# Make the prediction on the meshed x-axis
-y_pred = clf.predict(xx)
+# %%
+# This time, shallower trees are selected and lead to a more constant piecewise
+# and therefore more robust estimation of the 95th percentile. This is
+# beneficial as it avoids overfitting the large outliers of the log-normal
+# additive noise.
+#
+# We can confirm this intuition by displaying the predicted 90% confidence
+# interval comprised by the predictions of those two tuned quantile regressors:
+# the prediction of the upper 95th percentile has a much coarser shape than the
+# prediction of the lower 5th percentile:
```

Comment:
Should we compare the fraction of captured points of the first 90% interval and this fine-tuned one? This is a good measure for bias/calibration in this case.
We could also compare the test metrics.

Reply:
> Should we compare the fraction of captured points of the first 90% interval and this fine-tuned one? This is a good measure for bias/calibration in this case.
We could also compare the test metrics.

I thought about this but I was wondering if this would not add to much boilerplate. How would you name this quantity? `coverage_fraction`, `fraction_of_coverage`?

Reply:
I pushed a new commit where I assess calibration. Unfortunately, the tuned models are not better than the untuned ones. However the variability on this small data is bad. Feature engineering with splines + bagging would probably improve the quality of the 90% confidence interval estimate but I don't want to make this example any more complex.

Reply:
We optimized for validation/test error, not for calibration. If calibration is comparable, it's fine.